xbmc/utils/SysfsUtils.cpp

```diff
@@ -7,114 +7,85 @@
-  int ret = 0;
-  if (fd >= 0)
-  {
-    if (write(fd, valstr.c_str(), valstr.size()) < 0)
-      ret = -1;
-    close(fd);
-  }
-  if (ret)
-    CLog::Log(LOGERROR, "%s: error writing %s",__FUNCTION__, path.c_str());
-
-  return ret;
+const int PAGESIZE = getpagesize();
 }
 
-int SysfsUtils::GetString(const std::string& path, std::string& valstr)
+template<>
+std::string CSysfs::Get()
 {
-  int len;
-  char buf[256] = {0};
+  std::ifstream file(m_path);
 
-  int fd = open(path.c_str(), O_RDONLY);
-  if (fd >= 0)
-  {
-    valstr.clear();
-    while ((len = read(fd, buf, 256)) > 0)
-      valstr.append(buf, len);
-    close(fd);
+  if (!file.is_open())
+    return "";
 
-    StringUtils::Trim(valstr);
+  std::string value;
+  file >> value;
 
-    return 0;
+  if ((file.rdstate() & std::ifstream::badbit) != 0)
+  {
+    CLog::Log(LOGERROR, "CSysfs::{} error reading from '{}'", __FUNCTION__, m_path);
+    return "";
   }
 
-  CLog::Log(LOGERROR, "%s: error reading %s",__FUNCTION__, path.c_str());
-  valstr = "fail";
-  return -1;
+  return value;
 }
 
-int SysfsUtils::SetInt(const std::string& path, const int val)
+template<>
+int CSysfs::Get()
 {
-  int fd = open(path.c_str(), O_RDWR, 0644);
-  int ret = 0;
-  if (fd >= 0)
+  try
   {
-    char bcmd[16];
-    sprintf(bcmd, "%d", val);
-    if (write(fd, bcmd, strlen(bcmd)) < 0)
-      ret = -1;
-    close(fd);
+    return std::stoi(Get<std::string>());
+  }
+  catch (std::invalid_argument& error)
+  {
+    return -1;
   }
-  if (ret)
-    CLog::Log(LOGERROR, "%s: error writing %s",__FUNCTION__, path.c_str());
-
-  return ret;
 }
 
-int SysfsUtils::GetInt(const std::string& path, int& val)
+template<>
+bool CSysfs::Set(const std::string& value)
 {
-  int fd = open(path.c_str(), O_RDONLY);
-  int ret = 0;
-  if (fd >= 0)
+  if (Get<std::string>() == value)
+    return true;
+
+  if (static_cast<int>(value.size()) > PAGESIZE)
   {
-    char bcmd[16];
-    if (read(fd, bcmd, sizeof(bcmd)) < 0)
-      ret = -1;
-    else
-      val = strtol(bcmd, NULL, 16);
+    CLog::Log(LOGERROR, "CSysfs::{} value '{}' is larger then the system pagesize: {} vs {}",
+              __FUNCTION__, value, value.size(), PAGESIZE);
+    return false;
+  }
+
+  std::ofstream file(m_path);
 
-    close(fd);
+  if (!file.is_open())
+  {
+    CLog::Log(LOGERROR, "CSysfs::{} {} is not open for writing", __FUNCTION__, m_path);
+    return false;
   }
-  if (ret)
-    CLog::Log(LOGERROR, "%s: error reading %s",__FUNCTION__, path.c_str());
 
-  return ret;
-}
+  file << value;
 
-bool SysfsUtils::Has(const std::string &path)
-{
-  int fd = open(path.c_str(), O_RDONLY);
-  if (fd >= 0)
+  if ((file.rdstate() & std::ifstream::badbit) != 0)
   {
-    close(fd);
-    return true;
+    CLog::Log(LOGERROR, "CSysfs::{} error writing to '{}'", __FUNCTION__, m_path);
+    return false;
   }
-  return false;
+
+  return true;
 }
 
-bool SysfsUtils::HasRW(const std::string &path)
+template<>
+bool CSysfs::Set(const int& value)
```

Comment:
Having to take a const reference to an integer suggests that the templating strategy might not be optimal